File: 2022/04/java/Main.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.stream.Stream;

public class Main {
	private static final String INPUT_FILE = "../input/input.txt";

	/**
	 * Private class that represents a range.
	 */
	private static class Range {
		private final int min;
		private final int max;

		/**
		 * Constructor
		 *
		 * @param value String value to parse.
		 */
		private Range(String value) {
			String[] tokens = value.split("-");
			this.min = Integer.parseInt(tokens[0]);
			this.max = Integer.parseInt(tokens[1]);
		}

		/**
		 * Checks if this range fully contains another range.
		 *
		 * @param r Another range.
		 * @return A boolean
		 */
		private boolean fullyContains(Range r) {
			return this.min <= r.min && r.max <= this.max;
		}

		/**
		 * Checks if this range overlaps with another range.
		 *
		 * @param r Another range.
		 * @return A boolean
		 */
		private boolean overlap(Range r) {
			return
				(r.min >= this.min && r.min <= this.max) ||
				(r.max >= this.min && r.max <= this.max) ||
				(this.min >= r.min && this.min <= r.max) ||
				(this.max >= r.min && this.max <= r.max);
		}
	}

	public static void main(String[] args) throws IOException {
		// --- Puzzle 1 ---
		try (Stream<String> stream = Files.lines(Paths.get(INPUT_FILE))) {
			int total = stream.mapToInt(line -> {
				String[] tokens = line.split(",");
				Range r1 = new Range(tokens[0]);
				Range r2 = new Range(tokens[1]);
				return r1.fullyContains(r2) || r2.fullyContains(r1) ? 1 : 0;
			}).sum();
			System.out.format("Part 1. Total of ranges that fully contains another: %,d%n", total);
		}

		// -- Puzzle 2 ---
		try (Stream<String> stream = Files.lines(Paths.get(INPUT_FILE))) {
			int total = stream.mapToInt(line -> {
				String[] tokens = line.split(",");
				Range r1 = new Range(tokens[0]);
				Range r2 = new Range(tokens[1]);
				return r1.overlap(r2) ? 1 : 0;
			}).sum();
			System.out.format("Part 2. Total of ranges that overlaps: %,d%n", total);
		}
	}
}

File: 2022/04/rust/main.rs
use std::fs::File;
use std::io::{ BufRead, BufReader };
use std::num::ParseIntError;
use std::str::FromStr;

const INPUT_FILE:&'static str = "../input/input.txt";

// Private struct that represents a range.
#[derive(Debug)]
struct Range {
    min: usize,
    max: usize
}

impl Range {
    // Checks if this range fully contains another range.
    pub fn fully_contains(&self, r: &Range) -> bool {
        self.min <= r.min && r.max <= self.max
    }

    // Checks if this range overlaps with another range.
    pub fn overlap(&self, r: &Range) -> bool {
        (r.min >= self.min && r.min <= self.max) ||
        (r.max >= self.min && r.max <= self.max) ||
        (self.min >= r.min && self.min <= r.max) ||
        (self.max >= r.min && self.max <= r.max)
    }
}

impl FromStr for Range {
    type Err = ParseIntError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let i = s.find("-").unwrap();
        let min = s[0..i].parse::<usize>()?;
        let max = s[i + 1..].parse::<usize>()?;
        Ok(Range { min, max })
    }
}

fn main() -> Result<(), std::io::Error> {
    // --- Parse input file ---
    let file = File::open(INPUT_FILE)?;
    let list: Vec<(Range, Range)> = BufReader::new(file).lines().map(|line| {
        let line = line.unwrap();
        let i = line.find(",").unwrap();
        let r1 = line[0..i].parse::<Range>().unwrap();
        let r2 = line[i + 1..].parse::<Range>().unwrap();
        ( r1, r2 )
    }).collect();

    // --- Puzzle: Part 1 ---
    let total: usize = list.iter().map(|(r1, r2)| {
            if r1.fully_contains(r2) || r2.fully_contains(r1) { 1 } else { 0 }
        }).sum();
    println!("Part 1. Total of ranges that fully contains another: {}", total);

    // --- Puzzle: Part 2 ---
    let total: usize = list.iter().map(|(r1, r2)| {
            if r1.overlap(r2) { 1 } else { 0 }
        }).sum();
    println!("Part 2. Total of ranges that overlaps: {}", total);

    Ok(())
}
